### ISSUE TITLE
Upgrade taichi to use LLVM15

### ISSUE THREAD

**Issue by ailzhang** [taichi-dev/taichi#6447]
Thanks to @python3kgae we have pretty nice LLVM15 compatible code across the codebase now. And using an old version of LLVM (LLVM10) have limited us from applying advanced optimizations as well as adapting to newer hardware. So we're planning to upgrade taichi to depend on LLVM15. The overall process is kinda simple, fix the failed tests and enable TI_LLVM_15 in every CI job we have, and finally remove all the `ifdef TI_LLVM_15` in the codebase. 

A rough list of tasks:

- [x] linux cpu (#6435)
- [x] linux cuda (#6465)
- [x] mac cpu （#6634, #6483）
- [x] windows cpu
- [x] windows cuda (@python3kgae already did this, thanks!)
- [x] do we want to upgrade the compiler to compile `.bc` files? (#6678)
- [x] remove all the `ifdef TI_LLVM_15`! (#6685)
- [x] release prebuilt llvm binaries across different platforms & update docs (or shall we add llvm as a submodule?) (#6648)

**Comment by python3kgae** [comment]
> Thanks to @python3kgae we have pretty nice LLVM15 compatible code across the codebase now. And using an old version of LLVM (LLVM10) have limited us from applying advanced optimizations as well as adapting to newer hardware. So we're planning to upgrade taichi to depend on LLVM15. The overall process is kinda simple, fix the failed tests and enable TI_LLVM_15 in every CI job we have, and finally remove all the `ifdef TI_LLVM_15` in the codebase.
> 
> A rough list of tasks:
> 
> * [ ]  linux cpu ([[ci] Swtich CPU jobs to use prebuilt LLVM15 #6435](https://github.com/taichi-dev/taichi/pull/6435))
> * [ ]  linux cuda
> * [ ]  mac cpu
> * [ ]  windows cpu
> * [x]  windows cuda (@python3kgae already did this, thanks!)
> * [ ]  do we want to upgrade the compiler to compile `.bc` files?
> * [ ]  remove all the `ifdef TI_LLVM_15`!
> * [ ]  release prebuilt llvm binaries across different platforms & update docs (or shall we add llvm as a submodule?)

> * [ ]  windows cpu
windows cpu is enabled too.

> * [ ]  do we want to upgrade the compiler to compile `.bc` files?
It would be nice to upgrade clang to 15, that will generate more test cases for opaque ptr.

> * [ ]  release prebuilt llvm binaries across different platforms & update docs (or shall we add llvm as a submodule?)
rust add llvm as a submodule. But that will make clean build very slow.

**Comment by ailzhang** [comment]
> But that will make clean build very slow. Also clone llvm repo will take some time.
+1 that's the main reason we ship prebuilt llvm binaries on different platform IIRC. Building llvm from source does take a long time so we might have to stay on the current prebuilt approach ...

**Comment by ailzhang** [comment]
FYI #6649 turned on TI_LLVM_15 by default, we'll make sure nightly runs well and then remove all the `#ifdef TI_LLVM_15` in the codebase. 

Other tasks I'm looking into atm:

- [ ] Can we use llvm15 from brew on macs?  (for linux it's a bit tricky since llvm15 isn't available for all unix distributions yet)
- [ ] Can we upgrade compiler for `runtim.cpp` and handle it better?

**Comment by ailzhang** [comment]
FYI I'd consider LLVM15 upgrade is done at this point. There're two followups that might be helpful but with lower priority:

- [ ] Can we use llvm15 from brew on macs? (for linux it's a bit tricky since llvm15 isn't available for all unix distributions yet)
- [ ] Revamp developer install doc since it can be much simpler due to #6678 and #6682. 

I'll open separate issues to track these followups.
